service/sharddistributor/leader/process/processor.go

```diff
@@ -160,6 +177,56 @@ func (p *namespaceProcessor) runProcess(ctx context.Context) {
 	}
 }
 
+// runCleanupLoop periodically removes stale executors.
+func (p *namespaceProcessor) runCleanupLoop(ctx context.Context) {
+	ticker := p.timeSource.NewTicker(p.cfg.HeartbeatTTL)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			p.logger.Info("Cleanup loop cancelled.")
+			return
+		case <-ticker.Chan():
+			p.logger.Info("Periodic heartbeat cleanup triggered.")
+			p.cleanupStaleExecutors(ctx)
+		}
+	}
+}
+
+// cleanupStaleExecutors removes executors who have not reported a heartbeat recently.
+func (p *namespaceProcessor) cleanupStaleExecutors(ctx context.Context) {
+	// 1. Get the current heartbeat states. We don't need assignments for this operation.
+	heartbeatStates, _, _, err := p.shardStore.GetState(ctx)
+	if err != nil {
+		p.logger.Error("Failed to get state for heartbeat cleanup", tag.Error(err))
+		return
+	}
+
+	// 2. Identify expired executors.
+	var expiredExecutors []string
+	now := p.timeSource.Now().Unix()
+	heartbeatTTL := int64(p.cfg.HeartbeatTTL.Seconds())
+
+	for executorID, state := range heartbeatStates {
+		if (now - state.LastHeartbeat) > heartbeatTTL {
+			expiredExecutors = append(expiredExecutors, executorID)
+		}
+	}
+
+	if len(expiredExecutors) == 0 {
+		return // Nothing to do.
+	}
+
+	p.logger.Info("Removing stale executor", tag.ShardExecutors(expiredExecutors))
+	// 3. Remove the stale executors from the store.
+	for _, executorID := range expiredExecutors {
+		if err := p.shardStore.DeleteExecutor(ctx, executorID); err != nil {
+			p.logger.Error("Failed to delete stale executor", tag.Error(err), tag.ShardExecutor(executorID))
+		}
+	}
```

Comment:
This loop will notify the rebalance for each of the expired executors because it's not doing this in a signle transaction. Might be fine, but wanted to call it out

Reply:
Good point. We should not expect many of them to fail, but each deletion causes rebalancing, so it's preferable to do it in a transaction.